torch/csrc/autograd/python_function.cpp

```diff
@@ -146,35 +180,81 @@ auto PyNode::apply(variable_list&& inputs) -> variable_list {
-  for (int i = 0; i != num_outputs; ++i) {
-    PyObject* output = PyTuple_GET_ITEM(r.get(), i);
-    bool was_variable = is_variable_input[i];
-    if (!was_variable) {
-      if (output != Py_None) {
-        std::string msg("function ");
-        msg += name() + " returned a gradient different than None at position ";
-        msg += std::to_string(i + 1) +
-            ", but the corresponding forward input was not a Variable";
-        throw std::runtime_error(msg);
-      }
+  return to_variable_list(r.get(), is_variable_input);
+}
+
+auto PyNode::compiled_apply(
+    variable_list&& inputs,
+    std::optional<PyObject*> compiler) -> variable_list {
+  pybind11::gil_scoped_acquire gil;
+  at::OptionalDeviceGuard _device_guard;
+  THPFunction* py_fn = (THPFunction*)obj;
+
+  // Massage a C++ variable_list into a Python arguments tuple
+  THPObjectPtr pyInputs(to_py_args(inputs, &_device_guard));
+
+  const auto& is_variable_input = py_fn->is_variable_input;
+  const auto& input_infos = py_fn->input_info;
+  // input_info only contains info from variable inputs and should be a subset
+  TORCH_INTERNAL_ASSERT(is_variable_input.size() >= input_infos.size());
+
+  // The gradients returned in the backwards need to match the number of inputs
+  // to the forward, and their metadata, so we pass the fwdInputs
+  THPObjectPtr fwdInputMetadatas(
+      PyTuple_New(static_cast<Py_ssize_t>(is_variable_input.size())));
+  int offset = 0;
+  for (const auto i : c10::irange(is_variable_input.size())) {
+    if (!is_variable_input[i]) {
+      // input at i is not a variable, skip index
+      PyTuple_SET_ITEM(fwdInputMetadatas.get(), i, Py_None);
+      offset++;
       continue;
     }
-    if (output == Py_None) {
-      results.emplace_back();
-    } else {
-      if (!THPVariable_Check(output)) {
-        std::string msg("expected Variable or None (got ");
-        msg += THPUtils_typename(output);
-        msg += ")";
-        throw std::runtime_error(msg);
-      }
-      results.emplace_back(THPVariable_Unpack(output));
-    }
+
+    const auto& input_info = input_infos[i - offset];
+
+    // Metadata is a tuple of 4 elements: (layout, device, dtype, size)
+    PyObject* fwdInputMetadata(PyTuple_New(static_cast<Py_ssize_t>(4)));
+    PyTuple_SET_ITEM(
+        fwdInputMetadata,
+        static_cast<Py_ssize_t>(0),
+        autograd::utils::wrap(input_info.layout));
+    PyTuple_SET_ITEM(
+        fwdInputMetadata,
+        static_cast<Py_ssize_t>(1),
+        THPDevice_New(input_info.device));
+    PyTuple_SET_ITEM(
+        fwdInputMetadata,
+        static_cast<Py_ssize_t>(2),
+        autograd::utils::wrap(input_info.scalar_type));
+    PyTuple_SET_ITEM(
+        fwdInputMetadata,
+        static_cast<Py_ssize_t>(3),
+        to_py_size(input_info.size));
+
+    PyTuple_SetItem(
+        fwdInputMetadatas.get(), static_cast<Py_ssize_t>(i), fwdInputMetadata);
   }
+  PyObject* saved_tensors(unpack_saved_variables(
```

Comment:
I believe this object is getting leaked, since you own this reference.  Please double check object ownership to check for more leaks.

Every `PyObject*` should either be:
1) a borrowed reference
2) An owned reference you call Py_DECREF on to free (THPObjectPtr is doing this for you in other cases)
3) An owned reference you pass ownership to somewhere else (e.g. PyTuple_SetItem steals the referece)

Reply:
Updated to THPObjectPtr. This diff should only have 2 and 3